src/mono/mono/metadata/loader-internals.h

```diff
@@ -62,6 +72,53 @@ struct _MonoAssemblyLoadContext {
 };
 #endif /* ENABLE_NETCORE */
 
+struct _MonoMemoryManager {
+	// Whether the MemoryManager can be unloaded on netcore; should only be set at creation
+	gboolean collectible;
+	// Whether this is a singleton or generic MemoryManager
+	gboolean is_generic;
+	// Whether the MemoryManager is in the process of being freed
+	gboolean freeing;
+
+	// Entries moved over from the domain:
+
+	// If taking this with the loader lock, always take this second
+	// On legacy, this does _not_ protect mp/code_mp, which are covered by the domain lock
+	MonoCoopMutex lock;
+
+	MonoMemPool *mp;
+	MonoCodeManager *code_mp;
+
+	GPtrArray *class_vtable_array;
+
+	// !!! REGISTERED AS GC ROOTS !!!
+	// Hashtables for Reflection handles
+	MonoGHashTable *type_hash;
+	MonoConcGHashTable *refobject_hash;
+	// Maps class -> type initializaiton exception object
+	MonoGHashTable *type_init_exception_hash;
+	// Maps delegate trampoline addr -> delegate object
+	//MonoGHashTable *delegate_hash_table;
+	// End of GC roots
+};
+
+struct _MonoSingletonMemoryManager {
+	MonoMemoryManager memory_manager;
```

Comment:
Why there are two types of memory managers ?

Reply:
One is linked 1:1 with each ALC and one is used for generics. I don't introduce use of the generics one in this PR in the interest of breaking things up, since it's a rather large diff on its own.

Reply:
Does having two kinds makes things easier/faster ? I.e. having one alc owner is a special case of the other one.

Reply:
I think it's cleaner and easier to disambiguate. And for some purposes like cleanup, it's not actually just a special case of the other one. If I use just one struct I'll end up switching on an `is_generic` flag a bunch anyway, so it seems easier to use two structs, but this opinion isn't strongly held.